### ISSUE TITLE
IDN domains with valid umlauts should be converted to punycode

### ISSUE THREAD

**Issue by dakira** [swiftmailer/swiftmailer#541]
When an address like info@exämple.de is validated a Swift_RfcComplianceException is thrown, even though such domains are valid. When Swift_Mime_Grammar detects non-ascii chars in the domain part, it should just be converted to punycode by default. Lots of ~~idiots~~ uninformed people out there are using umlaut domains these days. ;(

**Comment by JustBlackBird** [comment]
+1

**Comment by starkj** [comment]
+1

concerning the 'idiots' that are using umlauts in domain names, have a look at this:

http://i.imgur.com/CK6aONG.jpg

time to make them all use swift mailer :-)

**Comment by gimler** [comment]
would be nice to have idn domains my actual workaround is:
``` php
$email = 'test@tüst.de';
list($name, $domain) = explode('@', $email);
$domain = idn_to_ascii($domain);
$email = $name . '@' . $domain; // test@xn--tst-hoa.de
```

`idn_to_ascii` is in the `intl` extension

**Comment by pmaselkowski** [comment]
@gimler As far as I know, `name` part should be converted too. This has to be made separatelly, or will yield bogus result, ie:
```php
$email = 'ł@msł.pl';
list($name, $domain) = explode('@', $email);
$domain = idn_to_ascii($domain);
$name = idn_to_ascii($name);

$result = $name . '@' . $domain;

// OK
// xn--tfa@xn--ms-yqa.pl
var_dump($result);

list($name, $domain) = explode('@', $result);

$domain = idn_to_utf8($domain);
$name = idn_to_utf8($name);

$result2 = $name . '@' . $domain;

// OK
// ł@msł.pl
var_dump($result2);

// Both wrong
// xn--@ms-gwad.pl - wrong conversion
var_dump(idn_to_ascii($email));
// xn--tfa@xn--ms-yqa.pl - not converted back to utf
var_dump(idn_to_utf8($result));

```

**Comment by voku** [comment]
@pmaselkowski 

I have implemented my old gist via "idn_to_ascii()" or "true/php-punycode" as fallback: https://github.com/voku/email-check/blob/master/src/voku/helper/EmailCheck.php#L790

And here are some valid e-mail addresses from the tests: https://github.com/voku/email-check/blob/master/tests/EmailCheckTest.php#L332 ;)

**Comment by ossinkine** [comment]
I've just published plugin for Swiftmailer which converts domain name to punycode before sending https://github.com/ossinkine/swiftmailer-punycode-plugin. Maybe someone will find it useful while we have no official support.

**Comment by c960657** [comment]
Fixed in #1044.

**Comment by dakira** [comment]
@c960657 thanks so much for tackling this!
